src/v/cloud_storage/remote_segment.cc

```diff
@@ -50,8 +50,51 @@
 
 #include <exception>
 
+namespace {
+class bounded_stream final : public ss::data_source_impl {
+public:
+    bounded_stream(ss::input_stream<char>& stream, size_t upto)
+      : _stream{stream}
+      , _upto{upto} {}
+
+    ss::future<ss::temporary_buffer<char>> get() override {
+        return _stream.read_exactly(_upto);
```

Comment:
nit: this is correct, but a more transparent adapter would probably use
```
auto buf=co_await _stream.read_up_to(_upto);
_upto-=buf.size();
co_return std::move(buf);
```
to not force allocation at this stage

Reply:
This appears to be why there were large allocations and subsequent failures on the latest commit, I have reverted the change not to use `read_exactly`.